torch/csrc/jit/frontend/ir_emitter.cpp

```diff
@@ -3004,6 +3005,164 @@ struct to_ir {
         return std::make_shared<SimpleValue>(
             emitListComprehension(lc, type_hint));
       }
+      case prim::dict: {
+        // Correctly set the type
+        TypePtr type;
+        if (type_hint) {
+          type = type_hint;
+          if (!type->cast<DictType>()) {
+            throw ErrorReport(apply.range())
+                << "Expected dict type annotation for dict(), found "
+                << type_hint->repr_str();
+          }
+        } else {
+          // If we have kwargs, get the type from the first kwarg
+          if (!apply.attributes().empty()) {
```

Comment:
Worth some discussion: Should we conduct more type checks to ensure passed key/values are of consistent types? ie. keys types can be unified and values types too.

Reply:
Hmm...I'll be shipping Union soon, which will change type unification. Should I wait to ship this until Union is out? I think it would be nice to allow for dicts with mixed keys just like Python does. For example, `{'a': 0, 1: 'b'}` would have a key type of `Union[str, int]`. (But adding checks is very easy so I don't mind doing that in the meantime.) What do you think?